Runtime/LLMChatHistory.cs

```diff
@@ -0,0 +1,147 @@
+/// @file
+/// @brief File implementing the LLMChatHistory.
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+
+namespace LLMUnity
+{
+    /// @ingroup llm
+    /// <summary>
+    /// Manages a single instance of a chat history.
+    /// </summary>
+    public class LLMChatHistory : MonoBehaviour
+    {
+        /// <summary> 
+        /// The name of the file where this chat history will be saved.
+        /// The file will be saved within the persistentDataPath directory (see https://docs.unity3d.com/ScriptReference/Application-persistentDataPath.html). 
+        /// </summary>
+        public string ChatHistoryFilename = string.Empty;
+
+        /// <summary>
+        /// If true, this component will automatically save a copy of its data to the filesystem with each update.
+        /// </summary>
+        public bool EnableAutoSave = true;
+
+        /// <summary>
+        /// The current chat history
+        /// </summary>
+        protected List<ChatMessage> _chatHistory;
+
+        /// <summary>
+        /// Ensures we're not trying to update the chat while saving or loading
+        /// </summary>
+        protected SemaphoreSlim chatLock = new SemaphoreSlim(1, 1);
+
+        /// <summary>
+        /// The Unity Awake function that initializes the state before the application starts.
+        /// </summary>
+        public async void Awake()
+        {
+            // If a filename has been provided for the chat history, attempt to load it
+            if (ChatHistoryFilename != string.Empty) {
+                await Load();
+            }
+            else {
+                _chatHistory = new List<ChatMessage>();
+            }
+        }
+
+        /// <summary>
+        /// Appends a new message to the end of this chat.
+        /// </summary>
+        public async Task AddMessage(string role, string content)
+        {
+            await WithChatLock(async () => {
+                await Task.Run(() => _chatHistory.Add(new ChatMessage { role = role, content = content }));
+            });
+
+            if (EnableAutoSave) {
+                _ = Save();
+            }
+        }
+
+        public List<ChatMessage> GetChatMessages() {
+            return new List<ChatMessage>(_chatHistory);
```

Comment:
Added the chat lock. Not returning directly because we don't want the actual _chatHistory list being edited directly by other objects (we want them to go through the methods defined in the LLMChatHistory). It does increase memory usage, that is true. I would argue it's probably not significant enough but that's your call.